File: apollo-integration/src/test/java/com/apollographql/apollo/CacheHeadersTest.java
package com.apollographql.apollo;

import android.support.annotation.NonNull;

import com.apollographql.apollo.api.Input;
import com.apollographql.apollo.cache.ApolloCacheHeaders;
import com.apollographql.apollo.cache.CacheHeaders;
import com.apollographql.apollo.cache.normalized.CacheKey;
import com.apollographql.apollo.cache.normalized.NormalizedCache;
import com.apollographql.apollo.cache.normalized.NormalizedCacheFactory;
import com.apollographql.apollo.cache.normalized.Record;
import com.apollographql.apollo.cache.normalized.RecordFieldJsonAdapter;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesQuery;
import com.apollographql.apollo.integration.normalizer.type.Episode;
import com.apollographql.apollo.rx2.Rx2Apollo;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.Collections;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import okhttp3.OkHttpClient;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static com.google.common.truth.Truth.assertThat;
import static java.util.Collections.emptySet;

public class CacheHeadersTest {
  private static final int TIME_OUT_SECONDS = 3;

  private MockWebServer server;

  @Before
  public void setUp() {
    server = new MockWebServer();
  }

  @After
  public void tearDown() throws IOException {
    server.shutdown();
  }

  @Test
  public void testHeadersReceived() throws ApolloException, IOException {
    final AtomicBoolean hasHeader = new AtomicBoolean();
    final NormalizedCache normalizedCache = new NormalizedCache() {
      @Nullable @Override public Record loadRecord(@NonNull String key, @NonNull CacheHeaders cacheHeaders) {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE));
        return null;
      }

      @Nonnull @Override public Set<String> merge(@NonNull Record record, @NonNull CacheHeaders cacheHeaders) {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE));
        return emptySet();
      }

      @Override public void clearAll() {
      }

      @Override public boolean remove(@Nonnull CacheKey cacheKey) {
        return false;
      }

      @Nonnull @Override
      protected Set<String> performMerge(@Nonnull Record apolloRecord, @Nonnull CacheHeaders cacheHeaders) {
        return emptySet();
      }
    };

    final NormalizedCacheFactory<NormalizedCache> cacheFactory = new NormalizedCacheFactory<NormalizedCache>() {
      @Override public NormalizedCache create(RecordFieldJsonAdapter recordFieldAdapter) {
        return normalizedCache;
      }
    };

    ApolloClient apolloClient = ApolloClient.builder()
        .normalizedCache(cacheFactory, new IdFieldCacheKeyResolver())
        .serverUrl(server.url("/"))
        .okHttpClient(new OkHttpClient())
        .build();

    server.enqueue(mockResponse("HeroAndFriendsNameResponse.json"));
    CacheHeaders cacheHeaders = CacheHeaders.builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "true").build();
    Rx2Apollo.from(apolloClient.query(new HeroAndFriendsNamesQuery(Input.fromNullable(Episode.NEWHOPE)))
        .cacheHeaders(cacheHeaders))
        .test().awaitDone(TIME_OUT_SECONDS, TimeUnit.SECONDS);
    assertThat(hasHeader.get()).isTrue();
  }

  @Test
  public void testDefaultHeadersReceived() throws IOException, ApolloException {
    final AtomicBoolean hasHeader = new AtomicBoolean();
    final NormalizedCache normalizedCache = new NormalizedCache() {
      @Nullable @Override public Record loadRecord(@NonNull String key, @NonNull CacheHeaders cacheHeaders) {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE));
        return null;
      }

      @Nonnull @Override public Set<String> merge(@NonNull Record record, @NonNull CacheHeaders cacheHeaders) {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE));
        return emptySet();
      }

      @Override public void clearAll() {
      }

      @Override public boolean remove(@Nonnull CacheKey cacheKey) {
        return false;
      }

      @Nonnull @Override
      protected Set<String> performMerge(@Nonnull Record apolloRecord, @Nonnull CacheHeaders cacheHeaders) {
        return emptySet();
      }
    };

    final NormalizedCacheFactory<NormalizedCache> cacheFactory = new NormalizedCacheFactory<NormalizedCache>() {
      @Override public NormalizedCache create(RecordFieldJsonAdapter recordFieldAdapter) {
        return normalizedCache;
      }
    };

    CacheHeaders cacheHeaders = CacheHeaders.builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "true").build();

    ApolloClient apolloClient = ApolloClient.builder()
        .normalizedCache(cacheFactory, new IdFieldCacheKeyResolver())
        .serverUrl(server.url("/"))
        .okHttpClient(new OkHttpClient())
        .defaultCacheHeaders(cacheHeaders)
        .build();

    server.enqueue(mockResponse("HeroAndFriendsNameResponse.json"));
    Rx2Apollo.from(apolloClient.query(new HeroAndFriendsNamesQuery(Input.fromNullable(Episode.NEWHOPE)))
        .cacheHeaders(cacheHeaders))
        .test()
        .awaitDone(TIME_OUT_SECONDS, TimeUnit.SECONDS);
    assertThat(hasHeader.get()).isTrue();
  }

  private MockResponse mockResponse(String fileName) throws IOException {
    return new MockResponse().setChunkedBody(Utils.readFileToString(getClass(), "/" + fileName), 32);
  }

}

File: apollo-integration/src/test/kotlin/com/apollographql/apollo/CacheHeadersTest.kt
package com.apollographql.apollo

import com.apollographql.apollo.Utils.immediateExecutor
import com.apollographql.apollo.Utils.immediateExecutorService
import com.apollographql.apollo.Utils.readFileToString
import com.apollographql.apollo.api.Input.Companion.fromNullable
import com.apollographql.apollo.api.internal.json.JsonReader
import com.apollographql.apollo.cache.ApolloCacheHeaders
import com.apollographql.apollo.cache.CacheHeaders
import com.apollographql.apollo.cache.CacheHeaders.Companion.builder
import com.apollographql.apollo.cache.normalized.*
import com.apollographql.apollo.exception.ApolloException
import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesQuery
import com.apollographql.apollo.integration.normalizer.type.Episode
import com.apollographql.apollo.rx2.Rx2Apollo
import com.google.common.truth.Truth
import okhttp3.Dispatcher
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.Rule
import org.junit.Test
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean

class CacheHeadersTest {
  val server = MockWebServer()
  @Test
  @Throws(ApolloException::class, IOException::class)
  fun testHeadersReceived() {
    val hasHeader = AtomicBoolean()
    val normalizedCache: NormalizedCache = object : NormalizedCache() {
      override fun loadRecord(key: String, cacheHeaders: CacheHeaders): Record? {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE))
        return null
      }

      override fun stream(key: String, cacheHeaders: CacheHeaders): JsonReader? {
        return null
      }

      override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE))
        return emptySet<String>()
      }

      override fun clearAll() {}
      override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
        return false
      }

      override fun performMerge(apolloRecord: Record, oldRecord: Record?, cacheHeaders: CacheHeaders): Set<String> {
        return emptySet()
      }
    }
    val cacheFactory: NormalizedCacheFactory<NormalizedCache> = object : NormalizedCacheFactory<NormalizedCache>() {
      override fun create(recordFieldAdapter: RecordFieldJsonAdapter): NormalizedCache {
        return normalizedCache
      }
    }
    val apolloClient = ApolloClient.builder()
        .normalizedCache(cacheFactory, IdFieldCacheKeyResolver())
        .serverUrl(server.url("/"))
        .okHttpClient(OkHttpClient.Builder().dispatcher(Dispatcher(immediateExecutorService())).build())
        .dispatcher(immediateExecutor())
        .build()
    server.enqueue(mockResponse("HeroAndFriendsNameResponse.json"))
    val cacheHeaders = builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "true").build()
    Rx2Apollo.from(apolloClient.query(HeroAndFriendsNamesQuery(fromNullable(Episode.NEWHOPE)))
        .cacheHeaders(cacheHeaders))
        .test()
    Truth.assertThat(hasHeader.get()).isTrue()
  }

  @Test
  @Throws(Exception::class)
  fun testDefaultHeadersReceived() {
    val hasHeader = AtomicBoolean()
    val normalizedCache: NormalizedCache = object : NormalizedCache() {
      override fun loadRecord(key: String, cacheHeaders: CacheHeaders): Record? {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE))
        return null
      }

      override fun stream(key: String, cacheHeaders: CacheHeaders): JsonReader? {
        return null
      }

      override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
        hasHeader.set(cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE))
        return emptySet<String>()
      }

      override fun clearAll() {}
      override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
        return false
      }

      override fun performMerge(apolloRecord: Record, oldRecord: Record?, cacheHeaders: CacheHeaders): Set<String> {
        return emptySet()
      }
    }
    val cacheFactory: NormalizedCacheFactory<NormalizedCache> = object : NormalizedCacheFactory<NormalizedCache>() {
      override fun create(recordFieldAdapter: RecordFieldJsonAdapter): NormalizedCache {
        return normalizedCache
      }
    }
    val cacheHeaders = builder().addHeader(ApolloCacheHeaders.DO_NOT_STORE, "true").build()
    val apolloClient = ApolloClient.builder()
        .normalizedCache(cacheFactory, IdFieldCacheKeyResolver())
        .serverUrl(server.url("/"))
        .okHttpClient(OkHttpClient.Builder().dispatcher(Dispatcher(immediateExecutorService())).build())
        .dispatcher(immediateExecutor())
        .defaultCacheHeaders(cacheHeaders)
        .build()
    server.enqueue(mockResponse("HeroAndFriendsNameResponse.json"))
    Rx2Apollo.from(apolloClient.query(HeroAndFriendsNamesQuery(fromNullable(Episode.NEWHOPE)))
        .cacheHeaders(cacheHeaders))
        .test()
    Truth.assertThat(hasHeader.get()).isTrue()
  }

  @Throws(IOException::class)
  private fun mockResponse(fileName: String): MockResponse {
    return MockResponse().setChunkedBody(readFileToString(javaClass, "/$fileName"), 32)
  }
}